### PR TITLE
Adding queue implementation core classes

### PR DESCRIPTION
**Pull request by bipinmaurya**
Adding autotune request queuing implementation code.

### CODE REVIEW

File: src/main/java/com/autotune/queue/ExpMgrQueue.java [hunk 1]
```diff
@@ -0,0 +1,72 @@
+/*******************************************************************************
```
Comment (dinogun) [inline src/main/java/com/autotune/queue/ExpMgrQueue.java:1]: I think the EM related Queue should ideally be inside the EM directory and should not be part of this PR. We just need a generic interface here that is not specific to any module and can be extended / implemented by the required modules.

File: src/main/java/com/autotune/queue/RecMgrQueue.java [hunk 2]
```diff
@@ -0,0 +1,72 @@
+/*******************************************************************************
```
Comment (dinogun) [inline src/main/java/com/autotune/queue/RecMgrQueue.java:1]: Same thing here as well

**Review by dinogun — CHANGES_REQUESTED** [review]

File: src/main/java/com/autotune/queue/AutotuneDTO.java [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package com.autotune.queue;
+
+import java.io.Serializable;
+
+import com.autotune.em.utils.EMUtils.EMProcessorType;
```
[the remark this answers is not in the document]
Comment (bipinmaurya) [reply src/main/java/com/autotune/queue/AutotuneDTO.java:20]: Removed the EMProcessorType from the AutotuneDTO

File: src/main/java/com/autotune/queue/AutotuneExecutor.java [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+package com.autotune.queue;
+
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Future;
+
+public class AutotuneExecutor {
+	
+	private final int DEFAULT_POOL_SIZE=5;
```
[the remark this answers is not in the document]
Comment (bipinmaurya) [reply src/main/java/com/autotune/queue/AutotuneExecutor.java:25]: fixed and documented.

File: src/main/java/com/autotune/queue/AutotuneDTO.java [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+
+
+public class AutotuneDTO implements Serializable {
+	private static final long serialVersionUID = 8789442223058487193L;
+	private int id;
+	private String name;
+	private String url;
+	private StringBuffer infoMessage;
+	private StringBuffer errorMessage;
+	private EMProcessorType processorName;
```
[the remark this answers is not in the document]
Comment (bipinmaurya) [reply src/main/java/com/autotune/queue/AutotuneDTO.java:30]: Removed the QueueProcessorType not needed because Shishir is not using this processor concept.

File: src/main/java/com/autotune/queue/AutotuneQueueFactory.java [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package com.autotune.queue;
+
+import com.autotune.em.utils.EMUtils.QueueName;
```
[the remark this answers is not in the document]
Comment (bipinmaurya) [reply src/main/java/com/autotune/queue/AutotuneQueueFactory.java:18]: referring from the AutotuneUtil.java

File: src/main/java/com/autotune/queue/ExpMgrQueue.java (hunk 1 above)
Comment (bipinmaurya) [reply src/main/java/com/autotune/queue/ExpMgrQueue.java:1]: AutotuneQueueFactory using both the instances so I have to put it as part of this PR if you want it in separate package then I have create and move this class.
Please let me know your decision?

File: src/main/java/com/autotune/queue/RecMgrQueue.java (hunk 2 above)
Comment (bipinmaurya) [reply src/main/java/com/autotune/queue/RecMgrQueue.java:1]: AutotuneQueueFactory using both the instances so I have to put it as part of this PR if you want it in separate package then I have create and move this class.
Please let me know your decision?

File: src/main/java/com/autotune/queue/AutotuneDTO.java [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+ * @author bipkumar
+ *
+ */
+
+public class AutotuneDTO implements Serializable {
+	private static final long serialVersionUID = 8789442223058487193L;
+	private int id;
+	private String name;
+	private String url;
+	private String operation;
```
Comment (dinogun) [inline src/main/java/com/autotune/queue/AutotuneDTO.java:31]: It looks like `operation` has been added here, but I dont see any getters and setters for it.

File: src/main/java/com/autotune/queue/AutotuneDTO.java [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	public String getUrl() {
+		return url;
+	}
+
+	public void setUrl(String url) {
+		this.url = url;
```
Comment (dinogun) [inline src/main/java/com/autotune/queue/AutotuneDTO.java:73]: As mentioned we need to do some error checking here to ensure that it is a valid URL

**Review by dinogun — CHANGES_REQUESTED** [review]

File: src/main/java/com/autotune/queue/ExpMgrQueue.java [hunk 9]
```diff
@@ -0,0 +1,60 @@
...
+
+import com.autotune.utils.AutotuneUtil;
+import com.autotune.utils.AutotuneUtil.QueueName;
+
+/**
+ * ExpMgrQueue is singleton concrete implementation of AutotuneQueue for Experiment Manager
+ *  * @author bipkumar
+ *
+ */
+public class ExpMgrQueue extends AutotuneQueueImpl implements Serializable {
```
Comment (dinogun) [inline src/main/java/com/autotune/queue/ExpMgrQueue.java:29]: Please name this as `ExperimentManagerQueue`

File: src/main/java/com/autotune/queue/RecMgrQueue.java [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+import java.util.concurrent.LinkedBlockingQueue;
+
+import com.autotune.utils.AutotuneUtil;
+import com.autotune.utils.AutotuneUtil.QueueName;
+/**
+ * RecMgrQueue is singleton concrete implementation of AutotuneQueue for Recommendation Manager.
+ * @author bipkumar
+ *
+ */
+public class RecMgrQueue extends AutotuneQueueImpl implements Serializable {
```
Comment (dinogun) [inline src/main/java/com/autotune/queue/RecMgrQueue.java:28]: Please name this as `RecommendationManagerQueue`

File: src/main/java/com/autotune/queue/AutotuneDTO.java [hunk 11]
```diff
@@ -0,0 +1,123 @@
...
+	public void setName(String name) {
+		this.name = name;
+	}
+
+	public String getUrl() {
+		return url;
+	}
+
+	public void setUrl(String url) {
+		if (url == null || url == "") {
```
Comment (dinogun) [inline src/main/java/com/autotune/queue/AutotuneDTO.java:82]: It appears that this check is already being done in `isValidURL`, so we dont need it here.

File: src/main/java/com/autotune/queue/EventProducer.java [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+	public Boolean call() {
+			try {
+				return queue.send(inputDTO);
+					
+			} catch (InterruptedException e) {
+				System.out.println("Adding Thread was interrupted");
+			} 
+			
+			return false;
+		}
```
Comment (dinogun) [inline src/main/java/com/autotune/queue/EventProducer.java:45]: Formatting seems to be off in this file

**Review by dinogun — CHANGES_REQUESTED** [review]

**Review by dinogun — APPROVED** [review]
lgtm
